File: utils/streakManager.ts
import { CheckIn } from '@/models/CheckIn';
import { SessionReport } from '@/models/Session';
import { StreakData } from '@/models/Streak';

class StreakManager {
  private static instance: StreakManager;

  private constructor() {}

  static getInstance(): StreakManager {
    if (!StreakManager.instance) {
      StreakManager.instance = new StreakManager();
    }
    return StreakManager.instance;
  }

  calculateStreak(
    checkIns: CheckIn[],
    sessionReports: SessionReport[],
    trainingDays: Record<string, string>
  ): StreakData {
    const activityDates = this.getUniqueActivityDates(checkIns, sessionReports);
    const trainingDayNames = new Set(Object.keys(trainingDays));

    const today = this.getStartOfDay(new Date());
    const todayName = this.getDayName(today);
    const isTodayTrainingDay = trainingDayNames.has(todayName);
    const completedToday = activityDates.has(this.formatDate(today));

    if (activityDates.size === 0) {
      return {
        currentStreak: 0,
        longestStreak: 0,
        lastActivityDate: undefined,
        isActive: !isTodayTrainingDay,
        completedToday: false,
        isTodayTrainingDay,
      };
    }

    const sortedActivityDates = Array.from(activityDates)
      .map((dateStr) => this.getStartOfDay(new Date(dateStr)))
      .sort((a, b) => b.getTime() - a.getTime());

    const mostRecentActivity = sortedActivityDates[0];
    if (!mostRecentActivity) {
      return {
        currentStreak: 0,
        longestStreak: 0,
        lastActivityDate: undefined,
        isActive: !isTodayTrainingDay,
        completedToday: false,
        isTodayTrainingDay,
      };
    }

    const isActive = this.checkStreakActive(
      today,
      mostRecentActivity,
      activityDates,
      trainingDayNames
    );

    const currentStreak = this.calculateCurrentStreak(
      sortedActivityDates,
      activityDates,
      trainingDayNames,
      today
    );

    const longestStreak = this.calculateLongestStreak(
      sortedActivityDates,
      activityDates,
      trainingDayNames
    );

    return {
      currentStreak,
      longestStreak: Math.max(longestStreak, currentStreak),
      lastActivityDate: this.formatDate(mostRecentActivity),
      isActive,
      completedToday,
      isTodayTrainingDay,
    };
  }

  private getUniqueActivityDates(
    checkIns: CheckIn[],
    sessionReports: SessionReport[]
  ): Set<string> {
    const dates = new Set<string>();

    checkIns.forEach((checkIn) => {
      const dateOnly = checkIn.check_in_date.substring(0, 10);
      dates.add(dateOnly);
    });

    sessionReports.forEach((report) => {
      const dateOnly = report.session_date.substring(0, 10);
      dates.add(dateOnly);
    });

    return dates;
  }

  private checkStreakActive(
    today: Date,
    mostRecentActivity: Date,
    activityDates: Set<string>,
    trainingDayNames: Set<string>
  ): boolean {
    const todayString = this.formatDate(today);
    if (activityDates.has(todayString)) {
      return true;
    }

    let checkDate = new Date(today);
    while (checkDate >= mostRecentActivity) {
      const checkDateString = this.formatDate(checkDate);
      const dayName = this.getDayName(checkDate);

      if (trainingDayNames.has(dayName)) {
        if (activityDates.has(checkDateString)) {
          return true;
        } else {
          return false;
        }
      }

      checkDate = new Date(checkDate.getTime() - 24 * 60 * 60 * 1000);
    }

    return true;
  }

  private calculateCurrentStreak(
    sortedActivityDates: Date[],
    activityDates: Set<string>,
    trainingDayNames: Set<string>,
    today: Date
  ): number {
    if (sortedActivityDates.length === 0) return 0;

    let streak = 0;
    let checkDate = new Date(today);
    let missedTrainingDay = false;

    const oldestActivity = sortedActivityDates[sortedActivityDates.length - 1] || today;
    const searchLimit = new Date(today.getTime() - 365 * 24 * 60 * 60 * 1000);

    while (checkDate >= searchLimit && !missedTrainingDay) {
      const checkDateString = this.formatDate(checkDate);
      const dayName = this.getDayName(checkDate);

      if (trainingDayNames.has(dayName)) {
        if (activityDates.has(checkDateString)) {
          streak += 1;
        } else {
          missedTrainingDay = true;
        }
      }

      checkDate = new Date(checkDate.getTime() - 24 * 60 * 60 * 1000);
    }

    return streak;
  }

  private calculateLongestStreak(
    sortedActivityDates: Date[],
    activityDates: Set<string>,
    trainingDayNames: Set<string>
  ): number {
    if (sortedActivityDates.length === 0) return 0;

    const oldestActivity = sortedActivityDates[sortedActivityDates.length - 1];
    const newestActivity = sortedActivityDates[0];

    if (!oldestActivity || !newestActivity) return 0;

    let longestStreak = 0;
    let currentStreak = 0;
    let checkDate = new Date(newestActivity);

    while (checkDate >= oldestActivity) {
      const checkDateString = this.formatDate(checkDate);
      const dayName = this.getDayName(checkDate);

      if (trainingDayNames.has(dayName)) {
        if (activityDates.has(checkDateString)) {
          currentStreak += 1;
          longestStreak = Math.max(longestStreak, currentStreak);
        } else {
          currentStreak = 0;
        }
      }

      checkDate = new Date(checkDate.getTime() - 24 * 60 * 60 * 1000);
    }

    return longestStreak;
  }

  private getStartOfDay(date: Date): Date {
    const newDate = new Date(date);
    newDate.setHours(0, 0, 0, 0);
    return newDate;
  }

  private formatDate(date: Date): string {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
  }

  private getDayName(date: Date): string {
    const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    return days[date.getDay()];
  }
}

export default StreakManager.getInstance();

File: swift-app/Utils/StreakManager.swift
//
//  StreakManager.swift
//  MeetJournal
//
//  Created by Maddisen Mohnsen on 1/11/26.
//

import Foundation

struct StreakData {
    var currentStreak: Int
    var longestStreak: Int
    var lastActivityDate: String?
    var isActive: Bool
    var daysUntilBreak: Int
    var completedToday: Bool
    var isTodayTrainingDay: Bool
}

class StreakManager {
    static let shared = StreakManager()
    
    private let calendar = Calendar.current
    private let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()
    
    private let dayNameFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE"
        return formatter
    }()
    
    private init() {}
    
    func calculateStreak(
        checkIns: [DailyCheckIn],
        sessionReports: [SessionReport],
        trainingDays: [String: String]
    ) -> StreakData {
        let activityDates = getUniqueActivityDates(checkIns: checkIns, sessionReports: sessionReports)
        let trainingDayNames = Set(trainingDays.keys)
        
        let today = calendar.startOfDay(for: Date())
        let todayName = dayNameFormatter.string(from: today)
        let isTodayTrainingDay = trainingDayNames.contains(todayName)
        let completedToday = activityDates.contains(dateFormatter.string(from: today))
        
        if activityDates.isEmpty {
            return StreakData(
                currentStreak: 0,
                longestStreak: 0,
                lastActivityDate: nil,
                isActive: !isTodayTrainingDay,
                daysUntilBreak: 0,
                completedToday: false,
                isTodayTrainingDay: isTodayTrainingDay
            )
        }
        
        let sortedActivityDates = activityDates
            .compactMap { dateFormatter.date(from: $0) }
            .map { calendar.startOfDay(for: $0) }
            .sorted(by: >)
        
        guard let mostRecentActivity = sortedActivityDates.first else {
            return StreakData(
                currentStreak: 0,
                longestStreak: 0,
                lastActivityDate: nil,
                isActive: !isTodayTrainingDay,
                daysUntilBreak: 0,
                completedToday: false,
                isTodayTrainingDay: isTodayTrainingDay
            )
        }
        
        let isActive = checkStreakActive(
            today: today,
            mostRecentActivity: mostRecentActivity,
            activityDates: activityDates,
            trainingDayNames: trainingDayNames
        )
        
        let currentStreak = calculateCurrentStreak(
            sortedActivityDates: sortedActivityDates,
            activityDates: activityDates,
            trainingDayNames: trainingDayNames,
            today: today
        )
        
        let longestStreak = calculateLongestStreak(
            sortedActivityDates: sortedActivityDates,
            activityDates: activityDates,
            trainingDayNames: trainingDayNames
        )
        
        return StreakData(
            currentStreak: currentStreak,
            longestStreak: max(longestStreak, currentStreak),
            lastActivityDate: dateFormatter.string(from: mostRecentActivity),
            isActive: isActive,
            daysUntilBreak: completedToday ? 1 : 0,
            completedToday: completedToday,
            isTodayTrainingDay: isTodayTrainingDay
        )
    }
    
    private func getUniqueActivityDates(checkIns: [DailyCheckIn], sessionReports: [SessionReport]) -> Set<String> {
        var dates = Set<String>()
        
        for checkIn in checkIns {
            let dateOnly = String(checkIn.check_in_date.prefix(10))
            dates.insert(dateOnly)
        }
        
        for report in sessionReports {
            let dateOnly = String(report.session_date.prefix(10))
            dates.insert(dateOnly)
        }
        
        return dates
    }
    
    private func checkStreakActive(
        today: Date,
        mostRecentActivity: Date,
        activityDates: Set<String>,
        trainingDayNames: Set<String>
    ) -> Bool {
        let todayString = dateFormatter.string(from: today)
        if activityDates.contains(todayString) {
            return true
        }
        
        var checkDate = today
        while checkDate >= mostRecentActivity {
            let checkDateString = dateFormatter.string(from: checkDate)
            let dayName = dayNameFormatter.string(from: checkDate)
            
            if trainingDayNames.contains(dayName) {
                if activityDates.contains(checkDateString) {
                    return true
                } else {
                    return false
                }
            }
            
            guard let previousDay = calendar.date(byAdding: .day, value: -1, to: checkDate) else {
                break
            }
            checkDate = previousDay
        }
        
        return true
    }
    
    private func calculateCurrentStreak(
        sortedActivityDates: [Date],
        activityDates: Set<String>,
        trainingDayNames: Set<String>,
        today: Date
    ) -> Int {
        guard !sortedActivityDates.isEmpty else { return 0 }
        
        var streak = 0
        var checkDate = today
        var missedTrainingDay = false
        
        let oldestActivity = sortedActivityDates.last ?? today
        let searchLimit = calendar.date(byAdding: .day, value: -365, to: today) ?? oldestActivity
        
        while checkDate >= searchLimit && !missedTrainingDay {
            let checkDateString = dateFormatter.string(from: checkDate)
            let dayName = dayNameFormatter.string(from: checkDate)
            
            if trainingDayNames.contains(dayName) {
                if activityDates.contains(checkDateString) {
                    streak += 1
                } else {
                    missedTrainingDay = true
                }
            }
            
            guard let previousDay = calendar.date(byAdding: .day, value: -1, to: checkDate) else {
                break
            }
            checkDate = previousDay
        }
        
        return streak
    }
    
    private func calculateLongestStreak(
        sortedActivityDates: [Date],
        activityDates: Set<String>,
        trainingDayNames: Set<String>
    ) -> Int {
        guard !sortedActivityDates.isEmpty,
              let oldestActivity = sortedActivityDates.last,
              let newestActivity = sortedActivityDates.first else {
            return 0
        }
        
        var longestStreak = 0
        var currentStreak = 0
        var checkDate = newestActivity
        
        while checkDate >= oldestActivity {
            let checkDateString = dateFormatter.string(from: checkDate)
            let dayName = dayNameFormatter.string(from: checkDate)
            
            if trainingDayNames.contains(dayName) {
                if activityDates.contains(checkDateString) {
                    currentStreak += 1
                    longestStreak = max(longestStreak, currentStreak)
                } else {
                    currentStreak = 0
                }
            }
            
            guard let previousDay = calendar.date(byAdding: .day, value: -1, to: checkDate) else {
                break
            }
            checkDate = previousDay
        }
        
        return longestStreak
    }
}
